File: 8-Linklist/programs/ruby/program1.rb
class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

def create(arr)
  first = Node.new(arr[0])
  last = first

  arr[1..-1].each do |value|
    t = Node.new(value)
    last.next = t
    last = t
  end

  first
end

def display(node)
  while node
    print "#{node.data} "
    node = node.next
  end
end

def recursiveDisplay(node)
  return if node.nil?

  print "#{node.data} "
  recursiveDisplay(node.next)
end

def recursiveDisplay1(node)
  return if node.nil?

  recursiveDisplay1(node.next)
  print "#{node.data} "
end

arr = [3, 5, 7, 10, 15]
first = create(arr)
display(first)
puts
recursiveDisplay(first)
puts
recursiveDisplay1(first)

File: 8-Linklist/programs/cpp/program1.cpp
// Link List

#include<bits/stdc++.h>
using namespace std;


struct Node{
  int data;
  struct Node *next;
}*first=NULL;


void create(int A[], int n){
  int i;
  struct Node *t, *last;
  first = new Node;
  first->data = A[0];
  first->next = NULL;
  last = first;

  for (i = 1; i < n; i++)
  {
    t = new Node;
    t->data = A[i];
    t->next = NULL;
    last->next = t;
    last = t;
  }
  
}

void Display(struct Node *p){
  while(p!=NULL){
    cout<< p->data << " ";
    p = p->next;
  }
}

void recursiveDisplay(struct Node *p){
  if(p!=NULL){
    cout<< p->data << " ";
    recursiveDisplay(p->next);
  }
}

void recursiveDisplay1(struct Node *p){
  if(p!=NULL){
    recursiveDisplay1(p->next);
    cout<< p->data << " ";
  }
}

int 

int main(){
  int A[] = {3,5,7,10,15};
  create(A,5);
  Display(first);
  cout << endl;
  recursiveDisplay(first);
  cout << endl;
  recursiveDisplay1(first);
  return 0;
}